File: Actito.iOS.Binding.Bridge/Sources/Models/ActitoApplication.swift
import Foundation
import ActitoKit

@objc
public class ActitoApplication : NSObject {
    @objc public let applicationId: String
    @objc public let name: String
    @objc public let category: String
    @objc public let appStoreId: String?
    @objc public let services: [String : Bool]
    @objc public let inboxConfig: ActitoApplicationInboxConfig?
    @objc public let regionConfig: ActitoApplicationRegionConfig?
    @objc public let userDataFields: [ActitoApplicationUserDataField]
    @objc public let actionCategories: [ActitoApplicationActionCategory]

    @objc public init(applicationId: String, name: String, category: String, appStoreId: String?, services: [String : Bool], inboxConfig: ActitoApplicationInboxConfig?, regionConfig: ActitoApplicationRegionConfig?, userDataFields: [ActitoApplicationUserDataField], actionCategories: [ActitoApplicationActionCategory]) {
        self.applicationId = applicationId
        self.name = name
        self.category = category
        self.appStoreId = appStoreId
        self.services = services
        self.inboxConfig = inboxConfig
        self.regionConfig = regionConfig
        self.userDataFields = userDataFields
        self.actionCategories = actionCategories
    }

    public convenience init(from application: ActitoKit.ActitoApplication) {
        self.init(
            applicationId: application.id,
            name: application.name,
            category: application.category,
            appStoreId: application.appStoreId,
            services: application.services,
            inboxConfig: application.inboxConfig.map { ActitoApplicationInboxConfig(from: $0) },
            regionConfig: application.regionConfig.map { ActitoApplicationRegionConfig(from: $0) },
            userDataFields: application.userDataFields.map { ActitoApplicationUserDataField(from: $0) },
            actionCategories: application.actionCategories.map { ActitoApplicationActionCategory(from: $0) }
        )
    }
}

@objc
public class ActitoApplicationInboxConfig : NSObject {
    @objc public let useInbox: Bool
    @objc public let useUserInbox: Bool
    @objc public let autoBadge: Bool

    @objc public init(useInbox: Bool, useUserInbox: Bool, autoBadge: Bool) {
        self.useInbox = useInbox
        self.useUserInbox = useUserInbox
        self.autoBadge = autoBadge
    }

    public convenience init(from inboxConfig: ActitoKit.ActitoApplication.InboxConfig) {
        self.init(
            useInbox: inboxConfig.useInbox,
            useUserInbox: inboxConfig.useUserInbox,
            autoBadge: inboxConfig.autoBadge
        )
    }
}

@objc
public class ActitoApplicationRegionConfig : NSObject {
    @objc public let proximityUUID: String?

    @objc public init(proximityUUID: String?) {
        self.proximityUUID = proximityUUID
    }

    public convenience init(from regionConfig: ActitoKit.ActitoApplication.RegionConfig) {
        self.init(proximityUUID: regionConfig.proximityUUID)
    }
}

@objc
public class ActitoApplicationUserDataField : NSObject {
    @objc public let type: String
    @objc public let key: String
    @objc public let label: String

    @objc public init(type: String, key: String, label: String) {
        self.type = type
        self.key = key
        self.label = label
    }

    public convenience init(from userDataField: ActitoKit.ActitoApplication.UserDataField) {
        self.init(
            type: userDataField.type,
            key: userDataField.key,
            label: userDataField.label
        )
    }
}

@objc
public class ActitoApplicationActionCategory : NSObject {
    @objc public let name: String
    @objc public let actionCategoryDescription: String?
    @objc public let type: String
    @objc public let actions: [ActitoNotificationAction]

    @objc public init(name: String, actionCategoryDescription: String?, type: String, actions: [ActitoNotificationAction]) {
        self.name = name
        self.actionCategoryDescription = actionCategoryDescription
        self.type = type
        self.actions = actions
    }

    public convenience init(from actionCategory: ActitoKit.ActitoApplication.ActionCategory) {
        self.init(
            name: actionCategory.name,
            actionCategoryDescription: actionCategory.description,
            type: actionCategory.type,
            actions: actionCategory.actions.map { ActitoNotificationAction(from: $0) }
        )
    }
}

File: Actito.Core/Models/ActitoApplication.cs
namespace ActitoSdk.Core.Models;

/// <summary>
/// Represents an Actito application.
/// </summary>
/// <remarks>
/// An <see cref="ActitoApplication"/> describes the capabilities, services, and configuration
/// of an application as defined in Actito. It includes enabled services, region and
/// inbox configuration, available user data fields, and supported action categories.
/// </remarks>
public class ActitoApplication
{
    /// <summary>
    /// Unique identifier of the application.
    /// </summary>
    public string Id { get; }

    /// <summary>
    /// Name of the application.
    /// </summary>
    public string Name { get; }

    /// <summary>
    /// Category of the application.
    /// </summary>
    public string Category { get; }

    /// <summary>
    /// The App Store ID of the application
    /// </summary>
    public string? AppStoreId { get; }

    /// <summary>
    /// Dictionary of enabled services for the application.
    /// </summary>
    public IDictionary<string, bool> Services { get; }

    /// <summary>
    /// Optional inbox-related configuration.
    /// </summary>
    public ActitoApplicationInboxConfig? InboxConfig { get; }

    /// <summary>
    /// Optional region-related configuration.
    /// </summary>
    public ActitoApplicationRegionConfig? RegionConfig { get; }

    /// <summary>
    /// List of user data fields supported by the application.
    /// </summary>
    public IList<ActitoApplicationUserDataField> UserDataFields { get; }

    /// <summary>
    /// List of action categories available in the application.
    /// </summary>
    public IList<ActitoApplicationActionCategory> ActionCategories { get; }

    /// <summary>
    /// Constructor for <see cref="ActitoApplication"/>.
    /// </summary>
    public ActitoApplication(string id, string name, string category, string? appStoreId, IDictionary<string, bool> services, ActitoApplicationInboxConfig? inboxConfig, ActitoApplicationRegionConfig? regionConfig, IList<ActitoApplicationUserDataField> userDataFields, IList<ActitoApplicationActionCategory> actionCategories)
    {
        Id = id;
        Name = name;
        Category = category;
        AppStoreId = appStoreId;
        Services = services;
        InboxConfig = inboxConfig;
        RegionConfig = regionConfig;
        UserDataFields = userDataFields;
        ActionCategories = actionCategories;
    }
}

/// <summary>
/// Configuration related to inbox-based features.
/// </summary>
public class ActitoApplicationInboxConfig
{
    /// <summary>
    /// Whether the inbox feature is enabled for the application.
    /// </summary>
    public bool UseInbox { get; }

    /// <summary>
    /// Whether the user inbox feature is enabled for the application.
    /// </summary>
    public bool UseUserInbox { get; }

    /// <summary>
    /// Whether inbox messages should automatically update the application badge count.
    /// </summary>
    public bool AutoBadge { get; }

    /// <summary>
    /// Constructor for <see cref="ActitoApplicationInboxConfig"/>.
    /// </summary>
    public ActitoApplicationInboxConfig(bool useInbox, bool useUserInbox, bool autoBadge)
    {
        UseInbox = useInbox;
        UseUserInbox = useUserInbox;
        AutoBadge = autoBadge;
    }
}

/// <summary>
/// Configuration related to region-based features.
/// </summary>
public class ActitoApplicationRegionConfig
{
    /// <summary>
    /// Optional UUID used for beacon detection.
    /// </summary>
    public string? ProximityUUID { get; }

    /// <summary>
    /// Constructor for <see cref="ActitoApplicationRegionConfig"/>.
    /// </summary>
    public ActitoApplicationRegionConfig(string? proximityUUID)
    {
        ProximityUUID = proximityUUID;
    }
}

/// <summary>
/// Describes a user data field supported by an Actito application.
/// </summary>
/// <remarks>
/// User data fields define the structure of user attributes that can be 
/// stored and leveraged for segmentation or personalization.
/// </remarks>
public class ActitoApplicationUserDataField
{
    /// <summary>
    /// The data type of the field.
    /// </summary>
    public string Type { get; }

    /// <summary>
    /// The unique key identifying the field.
    /// </summary>
    public string Key { get; }

    /// <summary>
    /// Human-readable label for the field.
    /// </summary>
    public string Label { get; }

    /// <summary>
    /// Constructor for <see cref="ActitoApplicationUserDataField"/>.
    /// </summary>
    public ActitoApplicationUserDataField(string type, string key, string label)
    {
        Type = type;
        Key = key;
        Label = label;
    }
}

/// <summary>
/// Groups related actions that can be triggered from notifications or other engagement mechanisms.
/// </summary>
public class ActitoApplicationActionCategory
{
    /// <summary>
    /// The category type identifier.
    /// </summary>
    public string Name { get; }

    /// <summary>
    /// The name of the action category.
    /// </summary>
    public string? Description { get; }

    /// <summary>
    /// Optional description explaining the purpose of the category.
    /// </summary>
    public string Type { get; }

    /// <summary>
    /// List of actions belonging to this category.
    /// </summary>
    public IList<ActitoNotificationAction> Actions { get; }

    /// <summary>
    /// Constructor for <see cref="ActitoApplicationActionCategory"/>.
    /// </summary>
    public ActitoApplicationActionCategory(string name, string? description, string type, IList<ActitoNotificationAction> actions)
    {
        Name = name;
        Description = description;
        Type = type;
        Actions = actions;
    }
}
